### ISSUE TITLE
GamePad.GetState() Crash in OS X

### ISSUE THREAD

**Issue by thejayjay** [mono/MonoGame#3790]
When spooling up a fresh project utilizing the 3.4.0455 build on OS X (which I acquired from TeamCity here -- http://teamcity.monogame.net/viewLog.html?buildId=6519&buildTypeId=MonoGame_DevelopMac&tab=artifacts) my project returns a null reference exception error -- as does the example project on this line

    if (GamePad.GetState (PlayerIndex.One) ...

Even simplifying to 
    var gamePad = GamePad.GetState (PlayerIndex.One);

causes the executable to crash with a null exception error.

`Unhandled Exception:
System.NullReferenceException: Object reference not set to an instance of an object
  at OpenTK.Input.GamePad.GetState (Int32 index) [0x00000] in <filename unknown>:0 
  at Microsoft.Xna.Framework.Input.GamePad.PlatformGetState (Int32 index, GamePadDeadZone deadZoneMode) [0x00000] in <filename unknown>:0 
  at Microsoft.Xna.Framework.Input.GamePad.GetState (PlayerIndex playerIndex, GamePadDeadZone deadZoneMode) [0x00000] in <filename unknown>:0 
  at Microsoft.Xna.Framework.Input.GamePad.GetState (PlayerIndex playerIndex) [0x00000] in <filename unknown>:0 
  at Toolbox.Game1.Update (Microsoft.Xna.Framework.GameTime gameTime) [0x00000] in <filename unknown>:0`

Xamarin Studio 5.8.3
Mono 3.12.1
GTK+ 2.24.23
XCode 6.2
OS X 10.9.5

3.3 assemblies worked fine.

Happy to provide any additional details if necessary.

**Comment by tomspilman** [comment]
I guess on OSX that is the OpenTK input path.  Right @cra0zy?

**Comment by cra0zy** [comment]
> I guess on OSX that is the OpenTK input path. Right @cra0zy?

Yup

> of an object at OpenTK.Input.GamePad.GetState (Int32 index) [0x00000]

Well it's due to OpenTK, at lease I know what I have to look into..

@thejayjay Could you try installing https://www.libsdl.org/download-2.0.php and try running the game agian

**Comment by thejayjay** [comment]
@cra0zy No luck -- just tried again, same null reference exception issue.  I placed the SDL2.framework in /Library/Frameworks.

**Comment by thefiddler** [comment]
@thejayjay hm, an exact stacktrace would help here. Copy [https://github.com/Mono-Game/MonoGame.Dependencies/blob/master/OpenTK.dll.mdb](OpenTK.dll.mdb) and [https://github.com/Mono-Game/MonoGame.Dependencies/blob/master/OpenTK.dll.config](OpenTK.dll.config) to your application directory, if they are not already there, and try running again. Post the resulting stacktrace here to help pinpoint the issue.

It might also be worth testing with the WindowsGL version of MonoGame (which sounds strange but does actually work for testing and is sometimes more stable.)

**Comment by thejayjay** [comment]
Took me awhile to confirm I had the .mdb in place; I'm usually in Visual Studio so this is all very new to me!

The stack trace still points directly at OpenTK.Input.GamePad.GetState; line 76.

https://github.com/opentk/opentk/blob/develop/Source/OpenTK/Input/GamePad.cs

If the source hasn't changed
`public static GamePadState GetState(int index)
{
    return driver.GetState(index);
}`

Longer stack trace
`
OpenTK.Input.GamePad.GetState (index=0) in /home/marko/GitHub/opentk/Source/OpenTK/Input/GamePad.cs:76
Microsoft.Xna.Framework.Input.GamePad.PlatformGetState (index=0, deadZoneMode=Microsoft.Xna.Framework.Input.GamePadDeadZone.IndependentAxes) in 
Microsoft.Xna.Framework.Input.GamePad.GetState (playerIndex=Microsoft.Xna.Framework.PlayerIndex.One, deadZoneMode=Microsoft.Xna.Framework.Input.GamePadDeadZone.IndependentAxes) in 
Microsoft.Xna.Framework.Input.GamePad.GetState (playerIndex=Microsoft.Xna.Framework.PlayerIndex.One) in 
TestSDLProject.Game1.Update (gameTime={Microsoft.Xna.Framework.GameTime}) in /Users/james/Projects/TestSDLProject/TestSDLProject/Game1.cs:63
`

Trying the WindowsGL assembly from the 3.4 visual studio installer yields app hanging results on the Game constructor.

Happy to try anything else -- including compiling OpenTK from source if the dependency rabbit hole isn't too deep.

**Comment by KonajuGames** [comment]
It would appear that driver is null.

**Comment by tomspilman** [comment]
@thefiddler - So is it right for `driver` to be null in `OpenTK.Input.GamePad`?  Or is this a bug within `Platform.Factory.Default.CreateGamePadDriver()`?

**Comment by cra0zy** [comment]
> Trying the WindowsGL assembly from the 3.4 visual studio installer yields app hanging results on the Game constructor.

WinGL works perfectly fine for me, what do you mean?

> So is it right for driver to be null in OpenTK.Input.GamePad? Or is this a bug within Platform.Factory.Default.CreateGamePadDriver()?

It should return a state that has IsConnected set to false, and with https://github.com/mono/MonoGame/blob/develop/MonoGame.Framework/Input/GamePad.OpenTK.cs#L86 it should get checked. It's probably a bug within mac implementation of Joystick.

Sorry if I am wrong about something.

**Comment by Frassle** [comment]
Copying comments from https://github.com/opentk/opentk/issues/257:

> Ok that's very suspicious as GetState is a simple call into driver.GetState so an NRE in that method must be due to driver being null. But then driver is created via a call to factory.CreateGamePadDriver which just returns a new MappedGamePadDriver() which,
A) Has a very simple ctor that I can't see how it could fail
B) Even if it did fail would cause an exception at that point, not later when it was used.

**Comment by tomspilman** [comment]
Are we sure the version of OpenTK we're using on MonoGame matches the current implementation?

**Comment by thejayjay** [comment]
I've created a template project leveraging the 3.4.0455 mac assemblies and replaced the OpenTK library with a version built from scratch.

After stepping through a lot of debugging in the platform factory for OpenTK I found the offending line that causes the factory initialization to abort.

`
Cocoa.SendVoid(Handle, Selector.Get("finishLaunching"));
`
https://github.com/opentk/opentk/blob/5953907f867d0fdecd83fda2aa31813041d9a514/Source/OpenTK/Platform/MacOS/Cocoa/NSApplication.cs#L87

The line above was returning this error:

`
_createMenuRef called with existing principal MenuRef already associated with menu
`

Google c/p debugging returned this answer.
http://stackoverflow.com/questions/7381345/cocoa-exception-createmenuref-called-with-existing-principal-menuref-already-a

I tried different source trees from OpenTK; the release from July 2014; the latest code in develop; and a branch from Jan 19 2015 -- which looking at the dependencies seems to be the version that might have been used to build.

The system even checks to make sure that mainMenu doesn't exist prior to menu creation; and the line does return 0x0; so I'm at a loss why the finish launching call bombs out.  Is there another way to check for a menu? Is it somehow named differently?

Removing the finish launching line does fix the issue, although I assume that's a hack.

Stack Trace for the menu bar crash
```
2015-05-02 12:50:43.086 TestProject[3995:e07] _createMenuRef called with existing principal MenuRef already associated with menu
2015-05-02 12:50:43.087 TestProject[3995:e07] (
	0   CoreFoundation                      0x9b4c3471 __raiseError + 193
	1   libobjc.A.dylib                     0x97cc9091 objc_exception_throw + 162
	2   CoreFoundation                      0x9b4c338b +[NSException raise:format:] + 139
	3   AppKit                              0x99d0bc9c -[NSCarbonMenuImpl _createMenuRef] + 69
	4   AppKit                              0x99d0b49d -[NSCarbonMenuImpl _instantiateCarbonMenu] + 161
	5   AppKit                              0x99d0b3e8 -[NSCarbonMenuImpl setupCarbonMenuBar] + 40
	6   AppKit                              0x99d098af -[NSApplication finishLaunching] + 1357
	7   ???                                 0x0a488584 0x0 + 172524932
	8   ???                                 0x0a4856dc 0x0 + 172512988
	9   ???                                 0x000f0602 0x0 + 984578
	10  libmono-2.0.dylib                   0x0030e4ff mono_jit_runtime_invoke + 751
	11  libmono-2.0.dylib                   0x004cd31f mono_runtime_invoke + 127
	12  libmono-2.0.dylib                   0x004cd84a mono_runtime_class_init_full + 778
	13  libmono-2.0.dylib                   0x0030b723 mono_jit_compile_method_with_opt + 3619
	14  libmono-2.0.dylib                   0x0030a8a9 mono_jit_compile_method + 57
	15  libmono-2.0.dylib                   0x003ba42f common_call_trampoline + 975
	16  libmono-2.0.dylib                   0x003ba04c mono_magic_trampoline + 60
	17  ???                                 0x000db088 0x0 + 897160
)
```

**Comment by Frassle** [comment]
> I tried different source trees from OpenTK; the release from July 2014; the latest code in develop; and a branch from Jan 19 2015 -- which looking at the dependencies seems to be the version that might have been used to build.

The check to see if a menu already exists and to not call finishLaunching if it does went into develop September 11 2014, so the July 2014 almost certainly has issues here.

> Is there another way to check for a menu? Is it somehow named differently?

I can't see anything in the OSX documentation, mainMenu is the name of property to get the application window.


Can you step though the initialization code and just check all the other calls we make to cocoa are returning sensible values, if one of the menu items is failing to allocate that might not be showing up till the finishLaunching call.

**Comment by thejayjay** [comment]
Everything looks sensible to me, unfortunately it's just a bunch of IntPtr's

Here's a screen shot of the locals.
![2015-05-04 20_49_54-jamess-macbook-pro local - chrome remote desktop](https://cloud.githubusercontent.com/assets/419952/7465585/b0378762-f2a0-11e4-950d-800162fe8db9.png)

The two items that are null are leveraged after the offending line; so I don't believe they are causing the issue.

I've also uploaded a directory tree that demonstrates the issue.  I left the only breakpoint that will cause the app to crash; you should be able to simply hit 'start debugging' and see the issue after stepping over the finishLaunching line.
https://dl.dropboxusercontent.com/u/47256/monogame-gamepad-issue/TestGamePad-monogame-opentk-3790.zip

**Comment by LithiumToast** [comment]
This problem can be easily solved by adding a MainMenu (MainMenu.xib) to the MacOS project and setting the interface in the project options (see screenshots).

<img width="1424" alt="screen shot 2015-08-07 at 2 46 45 pm" src="https://cloud.githubusercontent.com/assets/519592/9146645/4230d2ba-3d13-11e5-8e52-550674640cc4.png">

<img width="1424" alt="screen shot 2015-08-07 at 2 46 00 pm" src="https://cloud.githubusercontent.com/assets/519592/9146648/4748d07c-3d13-11e5-88f6-af610741843e.png">

This allows the method ```static NSApplication()`` in the ```OpenTK.Platform.MacOS.NSApplication``` class to find the menu and avoid entering the if statement for calling the offending line of code ```Cocoa.SendVoid(Handle, Selector.Get("finishLaunching"));```.

**Comment by cra0zy** [comment]
@dellis1972 If you have some time, could you look into adding MainMenu.xib to Mac templates, I tried but failed to find a solution.

**Comment by tomspilman** [comment]
Closing this...  either it is fixed in the 3.5 release or it will get replaced with the SDL2 change.
